scripts/spm-integration-test.sh

```diff
@@ -0,0 +1,109 @@
+#!/bin/bash
+
+set -e -uxf -o pipefail
+
+compose_file=docker-compose/monitor/docker-compose.yml
+
+dump_logs(){
+  echo "::group:: docker logs"
+  docker compose -f $compose_file logs
+  echo "::endgroup::"
+}
+
+# Function to check if a service is healthy
+check_service_health() {
+  local service_name=$1
+  local url=$2
+  local wait_seconds=5
+  local retry_count=10
+  echo "Checking health of service: $service_name at $url"
+  for i in $(seq 1 $retry_count); do
+    if curl -s -L --head --request GET "$url" | grep "200 OK" > /dev/null; then
+      echo "$service_name is healthy"
+      return 0
+    else
+      echo "Waiting for $service_name to be healthy... ($i/$retry_count)"
+      sleep $wait_seconds
+    fi
+  done
+
+  echo "Error: $service_name did not become healthy in time"
+  return 1
+}
+
+# Function to check if all services are healthy
+wait_for_services() {
+  echo "Waiting for services to be up and running..."
+  check_service_health "Jaeger" "http://localhost:16686"
+  check_service_health "Prometheus" "http://localhost:9090/graph"
+  check_service_health "Grafana" "http://localhost:3000"
+}
+
+# Function to check SPM
+check_spm() {
+  local timeout=180
+  local interval=5
+  local end_time=$((SECONDS + timeout))
+  echo "Checking SPM"
+  services_list=("driver" "customer" "mysql" "redis" "frontend" "route" "ui")
+  for service in "${services_list[@]}"; do
+    echo "Processing service: $service"
+    while [ $SECONDS -lt $end_time ]; do
+      response=$(curl -s "http://localhost:16686/api/metrics/calls?service=$service&endTs=$(date +%s)000&lookback=1000&step=100&ratePer=60000")
+      service_name=$(echo "$response" | jq -r 'if .metrics and .metrics[0] then .metrics[0].labels[] | select(.name=="service_name") | .value else empty end')
+      if [ "$service_name" != "$service" ]; then
+        echo "Service name does not match '$service'"
+        sleep $interval
```

Comment:
the logic here is very convoluted. I suggest you create a function that ONLY validates if the service metrics have the expected shape. That function can return false from multiple places, and only return true at the end.

Then in this code you only do the orchestration
```
for each service
  while not timeout
    if service_metrics_valid then
      successful_services++
      break
    else
      sleep 10
    fi
  end while
end for

if (( successful_services < len(services_list) )) then
  fail test
fi
```